### PR TITLE
fix: harden screen recording recovery

### PR DESCRIPTION
**Pull request by yjsoon**
## Summary

- serialise ScreenCaptureKit requests behind one process-wide recovery circuit for false `SCStreamErrorDomain/-3801` denials while TCC preflight remains granted
- escalate repeated cooldowns, surface persistent recovery guidance, and keep global recovery state authoritative during partial multi-display startup
- schedule and safely cancel/re-arm capture reconciliation after cooldown expiry
- add deterministic coverage for broker admission, escalation, cancellation, partial startup, and cooldown scheduler replacement

## Verification

- `xcodebuild test -project JustNow.xcodeproj -scheme JustNow -destination 'platform=macOS' -derivedDataPath /tmp/JustNow-pr-finalfull-b85c4e8 CODE_SIGNING_ALLOWED=NO CODE_SIGNING_REQUIRED=NO`
  - 273 tests passed, 0 failures on exact head `a4aca25`
- `git diff --check origin/main...HEAD`
- three independent final reviewers reported no actionable findings on exact head
- CodeRabbit status passed and every inline review thread is resolved
- the final signed reinstall was attempted through tmux, but macOS denied detached keychain signing with `errSecInternalComponent`; the previously installed Developer ID build remains valid, running, and capturing fresh frames

## Adversarial review

Three independent high-rigour Codex reviewers repeatedly inspected correctness/regressions, tests/types/edge cases, and lifecycle/operations as automated review added new findings. The resulting fixes cover reconciliation serialisation, generation-safe stop/start ordering, setup-time resume barriers, no-display fallback ownership, and multi-display readiness. The final exact head is clean across all three reviewers.

Related follow-up: #51 tracks strengthening the local installer signing-identity guard.

---

> [!NOTE]
> **High Risk**
> Changes core capture startup, permission-error handling, and async lifecycle ordering across ScreenCaptureKit; regressions could leave capture stuck, show wrong status, or miss real permission revocation.
> 
> **Overview**
> **Hardens screen recording recovery** when ScreenCaptureKit returns false permission denials (`-3801`) while TCC still shows access granted.
> 
> The **shared `CaptureRequestBroker` circuit** now **escalates cooldowns** (doubling up to a 10-minute cap), exposes **`needsAttention`** at max backoff, and **re-arms** after failed half-open probes without treating unrelated errors as healthy. **`CaptureCoordinator`** routes discovery through the broker, **serializes reconciliation** via a new gate, **schedules cooldown/fallback retries**, and only reports **healthy** after full multi-display reconciliation—not after an early probe.
> 
> **App lifecycle** treats broker cooldown and **no display** as **deferred** starts (not hard failures), shows **"Recovering…"** / **"Capture Help Needed"**, and presents a **one-time persistent recovery alert** with Screen Recording repair steps. **Start/stop ordering** waits for queued stops, tracks **deferred recovery** across overlay/session/launch setup, and avoids clearing deferred state when a newer resume wins.
> 
> **Tests** cover broker escalation, gate FIFO/cancellation, cooldown scheduler replacement, and deferred-start lifecycle edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a4aca25d61b91e25bf4f480ef8359a11dbff4e5d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved capture recovery when Screen Recording access temporarily fails or enters a cooldown period.
  * Capture attempts now retry automatically after recovery instead of showing misleading generic errors.
  * Added clearer statuses such as “Recovering…” and “Capture Help Needed.”
  * Detects permission revocation more reliably during capture startup and display changes.

* **New Features**
  * Added a persistent recovery alert with guidance for restoring Screen Recording permission.
  * Added a menu option to open the relevant system permission settings.
  * Improved handling of repeated failures with escalating recovery assistance.

### CODE REVIEW

**Comment by yjsoon** [comment]
@coderabbitai review

File: JustNow/AppDelegate.swift [hunk 1]
```diff
@@ -661,13 +685,22 @@ class AppDelegate: NSObject, NSApplicationDelegate, CaptureCoordinatorDelegate {
     ) {
         switch state {
         case .normal:
+            captureRecoveryNeedsAttention = false
+            updatePermissionHelpMenuItem()
```
[the remark this answers is not in the document]
Comment (yjsoon) [reply JustNow/AppDelegate.swift:739]: Intentional: the persistent modal is once per process to avoid repeated prompt storms during a prolonged macOS TCC desynchronisation. Recovery status and the Screen Recording help menu remain visible for later episodes, and relaunch resets the one-shot guard.

File: JustNow/Capture/CaptureRequestBroker.swift [hunk 2]
```diff
@@ -233,7 +311,7 @@ final class CaptureRequestBroker {
             "Global ScreenCaptureKit capture circuit \(event) for \(delay) seconds "
                 + "(queuedRequests=\(queuedCount), circuitOpenCount=\(openedCircuitCount))"
         )
-        recoveryStateDidChange(.coolingDown)
+        publishOpenCircuitRecoveryState()
```
[the remark this answers is not in the document]
Comment (yjsoon) [reply JustNow/Capture/CaptureRequestBroker.swift:308]: Fixed in the current head. Modal presentation now goes through schedulePersistentCaptureRecoveryAlert(), which enqueues a later @MainActor task. The broker finishes its synchronous turn, resets/releases its permit and resumes waiters before runModal() can execute; the task also re-checks that attention is still current.

File: JustNow/Capture/CaptureCoordinator.swift [hunk 3]
```diff
@@ -215,6 +362,63 @@ final class CaptureCoordinator: NSObject, ScreenCaptureDelegate {
...
+            do {
+                try await self.reconcileDisplays(startNewManagers: true)
+                guard self.isRunning, !Task.isCancelled else { return }
+                if self.isCapturing, self.captureRequestBroker.isCircuitClosed {
+                    // The broker publishes .normal when its half-open probe
+                    // succeeds, which happens before any display is capturing,
+                    // so the app skips the "Active" status update. Re-publish
+                    // now that capture is actually running again — but only if
+                    // no later display start reopened the circuit.
+                    self.delegate?.captureCoordinator(self, didChangeRecoveryState: .normal)
```
[the remark this answers is not in the document]
Comment (yjsoon) [reply JustNow/Capture/CaptureCoordinator.swift]: Fixed in the current head. Authoritative reconciled .normal now calls handleSuccessfulCaptureStart(), so background cooldown recovery restores capture policy/App Nap state as well as Active status. Readiness is gated on running + capturing + closed circuit.

File: JustNow/Capture/CaptureCoordinator.swift [hunk 4]
```diff
@@ -215,6 +362,63 @@ final class CaptureCoordinator: NSObject, ScreenCaptureDelegate {
...
+        DiagnosticsLog.shared.log(
+            "Capture",
+            "Capture start deferred; retrying in \(Int(delay)) seconds when the shared ScreenCaptureKit circuit cooldown ends"
+        )
+        cooldownRestartScheduler.schedule(
+            deadline: deadline,
+            delay: .seconds(delay)
+        ) { [weak self] in
+            guard let self else { return }
+            guard !Task.isCancelled, self.isRunning else { return }
```
[the remark this answers is not in the document]
Comment (yjsoon) [reply JustNow/Capture/CaptureCoordinator.swift:470]: Fixed in the current head. Deferred coordinator recovery is lifecycle-owned; overlay/session pause now treats pending recovery and launch setup as active, enqueues a coordinator stop, cancels the restart scheduler through stopCapture(), and preserves a later resume behind the ordered stop/setup barriers. Regression tests cover overlay setup/deferred recovery and stop ordering.
